File: examples/kotlin/src/main/kotlin/adr/blocks/escalation/Contract.kt
// ── blocks/escalation/contract — the block's transport ─────────────────────
// The gated slice: a REQUEST is reversible, a CONFIRM is not. Both are ordinary
// verbs with ordinary Command cases; the difference lives in the Verb table's
// reversibility classification and in the boundary gate, not in a second mechanic.

package adr.contract

import adr.spine.pure.CommandId
import adr.spine.pure.Signature
import adr.spine.pure.TicketId
import adr.spine.pure.Timestamp
import adr.spine.pure.ToolName

sealed interface EscalationResult : ToolResult {
    /**
     * L3 one level down: a BLOCK's sub-union declares its own shared property on its own
     * parent. Every escalation verb is about a ticket, so every variant carries one by
     * construction — and the block's fold arm never has to ask which case it has just to
     * find out which ticket it is talking about.
     */
    val ticket: TicketId

    data class RequestEscalation(
        override val tool: ToolName,
        override val ticket: TicketId,
    ) : EscalationResult

    data class ConfirmEscalation(
        override val tool: ToolName,
        override val ticket: TicketId,
    ) : EscalationResult
}

sealed interface EscalationCommand : Command {
    data class RequestEscalation(
        override val tool: ToolName,
        override val sig: Signature,
        override val id: CommandId,
        val ticket: TicketId,
    ) : EscalationCommand

    data class ConfirmEscalation(
        override val tool: ToolName,
        override val sig: Signature,
        override val id: CommandId,
        val ticket: TicketId,
    ) : EscalationCommand
}

sealed interface EscalationEffect : Effect {
    /** IRREVERSIBLE. Fires only inside the confirm arm's success branch. */
    data class PageOncall(
        override val at: Timestamp,
        val ticket: TicketId,
    ) : EscalationEffect
}

File: examples/typescript/src/blocks/escalation/contract.ts
// ── blocks/escalation/contract — the block's TRANSPORT (G12) ───────────────
// Two verbs, one reversible and one not; identical declarations otherwise.
// No Actor, no Authority, no Signature anywhere in this file (check C4) — the
// Command's `sig` is declared on the spine's parent, and a ToolResult never
// carries an actor at all, which is what makes one unforgeable upstream.

import type { CommandBase } from "@adr/spine/pure/command";
import type { EffectBase } from "@adr/spine/pure/effect";
import type { TicketId } from "@adr/spine/pure/ids";
import type { ToolResultBase } from "@adr/spine/pure/tool-result";

// ── ToolResult cases ────────────────────────────────────────────────────────
export interface RequestEscalationResult extends ToolResultBase {
  readonly outcome: "ok";
  readonly tool: "requestEscalation";
  readonly ticket: TicketId;
}

export interface ConfirmEscalationResult extends ToolResultBase {
  readonly outcome: "ok";
  readonly tool: "confirmEscalation";
  readonly ticket: TicketId;
}

export type EscalationResult = RequestEscalationResult | ConfirmEscalationResult;

// ── Command cases ───────────────────────────────────────────────────────────
export interface RequestEscalationCommand extends CommandBase {
  readonly outcome: "ok";
  readonly tool: "requestEscalation";
  readonly ticket: TicketId;
}

export interface ConfirmEscalationCommand extends CommandBase {
  readonly outcome: "ok";
  readonly tool: "confirmEscalation";
  readonly ticket: TicketId;
}

export type EscalationCommand = RequestEscalationCommand | ConfirmEscalationCommand;

// ── Effect cases ────────────────────────────────────────────────────────────
/** IRREVERSIBLE. It fires only from the success branch of the confirm arm, and
 *  only after the boundary gate admitted the confirm. */
export interface PageOncall extends EffectBase {
  readonly kind: "PageOncall";
  /** IRREVERSIBLE: a human is woken. Declared as a LITERAL, so no arm can spell
   *  this leaf `Routine` (docs/DECISIONS.md:85). */
  readonly effectClass: "Irreversible";
  readonly ticket: TicketId;
}

export type EscalationEffect = PageOncall;

export function isEscalationResult(r: ToolResultBase): r is EscalationResult {
  return r.outcome === "ok" && (r.tool === "requestEscalation" || r.tool === "confirmEscalation");
}
